### ISSUE TITLE
_TZ3000_wkai4ga5 actions interpreted twice and long press on buttons not supported

### ISSUE THREAD

**Issue by RBouroubi** [JohanBendz/com.tuya.zigbee#864]
### Describe the bug - //I've seen other issues posted but solutions aren't OK or users didn't give you the right information, I've decided to open a new one from scratch with (I hope) all necessary information.

> When we use 1 click or  2 clicks (with WHEN) the action is done twice (same for all 4 buttons)
>The "Long press" action isn't listed

### To Reproduce

I've purshased new ones, I did a simple Alternate on click and the lights are going ON and OFF again directly.
I created an advanced flow with variables to increment when I click or double click and each time an action is done once, the variable increments twice.

### Expected behavior

> Only one action 
> Add "Long Click" on the actions list

### App and Homey version

> Homey Pro (Early 2019)
> 10.0.9
> Tuya Zigbee App v0.2.26 dev

### Additional context

> Please find the Interview of the device :

  "ids": {
    "modelId": "TS0044",
    "manufacturerName": "_TZ3000_wkai4ga5"
  },
  "endpoints": {
    "endpointDescriptors": [
      {
        "endpointId": 1,
        "applicationProfileId": 260,
        "applicationDeviceId": 0,
        "applicationDeviceVersion": 0,
        "_reserved1": 1,
        "inputClusters": [
          0,
          1,
          6
        ],
        "outputClusters": [
          25,
          10
        ]
      },
      {
        "endpointId": 2,
        "applicationProfileId": 260,
        "applicationDeviceId": 0,
        "applicationDeviceVersion": 0,
        "_reserved1": 1,
        "inputClusters": [
          1,
          6
        ],
        "outputClusters": []
      },
      {
        "endpointId": 3,
        "applicationProfileId": 260,
        "applicationDeviceId": 0,
        "applicationDeviceVersion": 0,
        "_reserved1": 1,
        "inputClusters": [
          1,
          6
        ],
        "outputClusters": []
      },
      {
        "endpointId": 4,
        "applicationProfileId": 260,
        "applicationDeviceId": 0,
        "applicationDeviceVersion": 0,
        "_reserved1": 1,
        "inputClusters": [
          1,
          6
        ],
        "outputClusters": []
      }
    ],
    "endpoints": {
      "1": {
        "clusters": {
          "basic": {},
          "powerConfiguration": {},
          "onOff": {}
        },
        "bindings": {
          "ota": {
            "attributes": [],
            "commandsGenerated": "UNSUP_GENERAL_COMMAND",
            "commandsReceived": "UNSUP_GENERAL_COMMAND"
          },
          "time": {
            "attributes": [],
            "commandsGenerated": "UNSUP_GENERAL_COMMAND",
            "commandsReceived": "UNSUP_GENERAL_COMMAND"
          }
        }
      },
      "2": {
        "clusters": {
          "powerConfiguration": {
            "attributes": [
              {
                "acl": [
                  "readable",
                  "reportable"
                ],
                "id": 0,
                "reportingConfiguration": {
                  "status": "NOT_FOUND",
                  "direction": "reported"
                }
              },
              {
                "acl": [
                  "readable",
                  "reportable"
                ],
                "id": 32,
                "name": "batteryVoltage",
                "value": 30,
                "reportingConfiguration": {
                  "status": "NOT_FOUND",
                  "direction": "reported"
                }
              },
              {
                "acl": [
                  "readable",
                  "reportable"
                ],
                "id": 33,
                "name": "batteryPercentageRemaining",
                "value": 200,
                "reportingConfiguration": {
                  "status": "NOT_FOUND",
                  "direction": "reported"
                }
              },
              {
                "acl": [
                  "readable",
                  "reportable"
                ],
                "id": 65533,
                "name": "clusterRevision",
                "value": 1,
                "reportingConfiguration": {
                  "status": "NOT_FOUND",
                  "direction": "reported"
                }
              }
            ]
          },
          "onOff": {}
        },
        "bindings": {}
      },
      "3": {
        "clusters": {
          "powerConfiguration": {},
          "onOff": {}
        },
        "bindings": {}
      },
      "4": {
        "clusters": {
          "powerConfiguration": {},
          "onOff": {}
        },
        "bindings": {}
      }
    }
  }

**Comment by roxart** [comment]
is this hard to fix Johan? Can I somehow help reviewing the according code to see where a double trigger could come from?

**Comment by RBouroubi** [comment]
Hi,
I have the devices, if you need to perform any test, please let me know.

**Comment by JohanBendz** [comment]
Hi @RBouroubi and @roxart.
Can you please be very specific about what does not work and what is missing?
This is what the driver currently support:

Upper Left Button: 1 Click
Upper Left Button: 2 Clicks
Upper Right Button: 1 Click
Upper Right Button: 2 Clicks
Lower Left Button: 1 Click
Lower Left Button: 2 Clicks
Lower Right Button: 1 Click
Lower Right Button: 2 Clicks

Are these working? If not, what is wrong..
Is "longpress" missing?

If so please use the buttons in all ways and then send me a diagnostic report from the app.

**Comment by RBouroubi** [comment]
Hi Johan,
The long press is missing.
When you click once, two signals are coming to Homey.
For example, if I create a workflow to alternate ON and OFF for a light, it
doesn't alternate because it goes on and off very quickly.
I put a variable to count how many times the switch sends the command, and
it's always twice.
I don't know how to send a diagnostic report, I've attached in the first
post the discovery JSON.

**Comment by roxart** [comment]
![{D1C3A92A-949D-44E6-8A31-2B36D4A9D541}](https://github.com/user-attachments/assets/20f51bac-4b82-46b5-bb50-4900657169ab)
Yes Johan it works - but every thing triggers 2 times - so the example above gives me 2 push notifications

and there is no longpress correct

this is the report right after I clicked with that flow on: 7a38ea4c-8d33-44c2-aa79-fbf96f57fb0d

**Comment by roxart** [comment]
also I want to add: the longpress is actually important for me, as I want to use some of these buttons to control blinds and longpress is to go up and stop on release
what is the problem, that those do not work?

**Comment by JohanBendz** [comment]
@roxart An update to the driver is on it's way, please try and report back. If anything is not working please send a diagnostic report.

**Comment by SunBeech** [comment]
In the current test version of the app (v0.2.68) the 4 button switch does not work anymore. I understand (based on your post in the Homey Community Forum) that you are working in a new test version. Thanks in advance for all your efforts. Is there any information I can provide to help you further identify / test the issue?

I did note something else that might be relevant:

- The issues with a button press registered twice also relates to the 2 scene (and possible other) versions of this device. In my case, I have the MOES 2 button version: Product ID TS0042, Manufacturer ID: _TZ3000_dfgbtub0.
- However, this 2 scene MOES device uses the same device ID as the equivalent device from Marmitek. If I pair this switch to the Marmitek app (which supports 1 and 2 button versions of the switch), it works flawlessly: presses are registered only once (and both the single, double, and long press options are available).

If I am not mistaken, you are also involved / contributed to the Marmitek app. Maybe the code of that apps might help in pintointing the issue.

**Comment by JohanBendz** [comment]
Hi @SunBeech, thanks for this information.. always hard to code for device i don't have so I rely on community testing :)
OK, good info.. and what is your id for the 4 button version?
_TZ3000_wkai4ga5, _TZ3000_uaa99arv, _TZ3000_jcspr0tp

Yeah, I code the Marmitek app too.. hmm.. just had a look at that code and *darn*.. you are right... I made a misstake in the Tuya code. I just updated the 2 button code in the tuya app, try it out and let me know if this works better. (v.0.2.71)

**Comment by roxart** [comment]
for me it still triggers twice on the 4 gang @JohanBendz

**Comment by SunBeech** [comment]
Hi @JohanBendz, regarding the 4 button version. The ID is: _TZ3000_wkai4ga5.

In the new test version (v0.2.71) I can confirm @roxart's observation: it triggers twice. 
I tested all combinations of button presses: it always triggers twice, except for one case, in which it does not trigger at all (lower right long press)

1 click
Upper Left: triggers twice
Upper Right: triggers twice
Lower Left: triggers twice
Lower Right: triggers twice

2 clicks
Upper Left: triggers twice
Upper Right: triggers twice
Lower Left: triggers twice
Lower Right: triggers twice

Long press
Upper Left: triggers twice
Upper Right: triggers twice
Lower Left: triggers twice
**Lower Right: NO TRIGGER**

I'll get back to you on the 2 button device.

**Comment by SunBeech** [comment]
Hi @JohanBendz, I checked the 2 button switch (_TZ3000_dfgbtub0), which works with both the Tuya Zigbee app and the Marmitek app.

I first checked that the device is still working as expected on the current version of the Marmitek app (v1.1.0, as that was also recently updated): all combinations of button presses work as expected and trigger only once.

I then removed the device from my Homey and added it back again using the Tuya Zigbee app (v0.2.71). Also now the buttons work as expected: a single, double or long press on any of the buttons result in a single trigger only.

The only minor thing I noted relates to the names of the button actions in the WHEN card:
![Tuya 2 Gang Button Action Names](https://github.com/user-attachments/assets/7814d20f-da67-4a01-9604-0ff104c1f119)
In says 'right button was clicked 2 times' while in all other names the word 'pressed' is used instead of 'clicked'.

The 4 button version uses different namings, though they are consistent :-).
![Tuya 4 Gang Button Action Names](https://github.com/user-attachments/assets/3853fae8-3dae-4bc1-9695-3bc608da3441)

**Comment by JohanBendz** [comment]
> Also now the buttons work as expected: a single, double or long press on any of the buttons result in a single trigger only.

Sweet! Thanks for verifying @SunBeech 👍

Ahhh.. naming standard.. I'll make it consistent.. if only in the app 😄

@roxart And now I can fix the other buttons.. an update is coming.

**Comment by RBouroubi** [comment]
Hi @JohanBendz ,
I'm struggling to install this version, when I make the install i have messages on the validation step.
But As I installed only once fe weeks ago, maybe I'm doing something wrong

```
homey app install
✓ Pre-processing app...
Added FlowCard `the-breath-alarm-turned-off` for type `triggers`
Added FlowCard `the-breath-alarm-turned-on` for type `triggers`
Added FlowCard `the-large-presence-alarm-turned-off` for type `triggers`
Added FlowCard `the-large-presence-alarm-turned-on` for type `triggers`
Added FlowCard `the-small-presence-alarm-turned-off` for type `triggers`
Added FlowCard `the-small-presence-alarm-turned-on` for type `triggers`
Added FlowCard `the-breath-alarm-is-onoff` for type `conditions`
Added FlowCard `the-large-presence-alarm-is-onoff` for type `conditions`
Added FlowCard `the-small-presence-alarm-is-onoff` for type `conditions`
Added FlowCard `resetEnergyMeter` for type `actions`
Added FlowCard `start_effect` for type `actions`
Added Driver `christmas_lights`
Added FlowCard `move_open` for type `actions`
Added FlowCard `move_close` for type `actions`
Added Driver `curtain_module`
Added Driver `curtain_motor`
Added Driver `dimmer_1_gang`
Added Driver `dimmer_1_gang_tuya`
Added Driver `dimmer_2_gang`
Added Driver `dimmer_2_gang_tuya`
Added Driver `doorwindowsensor`
Added Driver `doorwindowsensor_2`
Added Driver `doorwindowsensor_3`
Added Driver `double_power_point`
Added Driver `flood_sensor`
Added FlowCard `handheld_remote_4_buttons` for type `triggers`
Added Driver `handheld_remote_4_buttons`
Added Driver `lcdtemphumidluxsensor`
Added Driver `lcdtemphumidsensor`
Added Driver `lcdtemphumidsensor_2`
Added Driver `lcdtemphumidsensor_3`
Added Driver `motion_sensor`
Added Driver `motion_sensor_2`
Added Driver `outdoor_2_socket`
Added Driver `outdoor_plug`
Added Driver `pir_sensor_2`
Added Driver `pirsensor`
Added Driver `plug`
Added Driver `radar_sensor`
Added Driver `radar_sensor_2`
Added Driver `relay_board_4_channel`
Added Driver `rgb_bulb_E14`
Added Driver `rgb_bulb_E27`
Added Driver `rgb_led_light_bar`
Added Driver `rgb_led_strip`
Added Driver `rgb_led_strip_controller`
Added Driver `rgb_mood_light`
Added Driver `rgb_spot_GU10`
Added Driver `rgb_spot_GardenLight`
Added FlowCard `alarm_siren` for type `triggers`
Added FlowCard `alarm_state` for type `actions`
Added FlowCard `alarm_duration` for type `actions`
Added FlowCard `siren_volume` for type `actions`
Added FlowCard `alarm_tune` for type `actions`
Added Driver `sirentemphumidsensor`
Added Driver `slim_motion_sensor`
Added Driver `smart_door_window_sensor`
Added Driver `smart_garden_irrigation_control`
Added Driver `smart_motion_sensor`
Added FlowCard `smart_remote_1_button` for type `triggers`
Added Driver `smart_remote_1_button`
Added FlowCard `smart_remote_1_button_2` for type `triggers`
Added Driver `smart_remote_1_button_2`
Added FlowCard `smart_remote_4_buttons` for type `triggers`
Added Driver `smart_remote_4_buttons`
Added Driver `smart_switch`
Added Driver `smart_water_timer`
Added Driver `smartplug`
Added Driver `smartplug_2_socket`
Added Driver `smoke_sensor`
Added Driver `smoke_sensor2`
Added Driver `socket_power_strip`
Added Driver `socket_power_strip_four`
Added Driver `socket_power_strip_four_three`
Added Driver `socket_power_strip_four_two`
Added Driver `soilsensor`
Added Driver `switch_1_gang`
Added Driver `switch_2_gang`
Added Driver `switch_2_gang_metering`
Added Driver `switch_3_gang`
Added Driver `temphumidsensor`
Added Driver `temphumidsensor2`
Added Driver `temphumidsensor3`
Added FlowCard `window_open_status_get` for type `conditions`
Added FlowCard `window_open_status_set` for type `actions`
Added Driver `thermostatic_radiator_valve`
Added Driver `tunable_bulb_E14`
Added Driver `tunable_bulb_E27`
Added Driver `tunable_spot_GU10`
Added Driver `valvecontroller`
Added Driver `wall_curtain_switch`
Added Driver `wall_dimmer_tuya`
Added FlowCard `wall_remote_1_gang_buttons` for type `triggers`
Added Driver `wall_remote_1_gang`
Added FlowCard `wall_remote_2_gang_buttons` for type `triggers`
Added Driver `wall_remote_2_gang`
Added FlowCard `wall_remote_3_gang_buttons` for type `triggers`
Added Driver `wall_remote_3_gang`
Added FlowCard `wall_remote_4_gang_buttons` for type `triggers`
Added Driver `wall_remote_4_gang`
Added FlowCard `wall_remote_4_gang_buttons_2` for type `triggers`
Added Driver `wall_remote_4_gang_2`
Added FlowCard `wall_remote_4_gang_buttons_3` for type `triggers`
Added Driver `wall_remote_4_gang_3`
Added Driver `wall_switch_1_gang`
Added Driver `wall_switch_1_gang_tuya`
Added Driver `wall_switch_2_gang`
Added Driver `wall_switch_3_gang`
Added Driver `wall_switch_4_gang`
Added Driver `wall_thermostat`
Added Driver `water_detector`
Added Driver `water_leak_sensor_tuya`
Added Driver `zigbee_repeater`
Added Capability `alarm_breathe_presence`
Added Capability `alarm_large_presence`
Added Capability `alarm_siren`
Added Capability `alarm_small_presence`
Added Capability `away_mode`
Added Capability `child_lock`
Added Capability `last_watering_time`
Added Capability `lidl_xmas_mode`
Added Capability `measure_water_flow`
Added Capability `meter_valve_state`
Added Capability `remaining_watering_time`
Added Capability `shutdown_timer`
Added Capability `soil_moisture`
Added Capability `target_distance`
Added Capability `temp_calibration`
Added Capability `temperature_unit`
Added Capability `thermostat_preset`
Added Capability `valve_state_auto_shutdown`
Added Capability `window_open`
✓ Validating app...
× App did not validate against level `debug`:
× Error: manifest.drivers['curtain_module'].settings[0].type should be equal to one of the allowed values ["text","password","textarea","label"]
manifest.drivers['curtain_module'].settings[0] should have required property 'id'
manifest.drivers['curtain_module'].settings[0].type should be equal to one of the allowed values ["number","slider"]
manifest.drivers['curtain_module'].settings[0] should have required property 'id'
manifest.drivers['curtain_module'].settings[0].type should be equal to one of the allowed values ["radio","dropdown"]
manifest.drivers['curtain_module'].settings[0] should have required property 'id'
manifest.drivers['curtain_module'].settings[0] should have required property 'values'
manifest.drivers['curtain_module'].settings[0].type should be equal to one of the allowed values ["checkbox"]
manifest.drivers['curtain_module'].settings[0] should have required property 'id'
manifest.drivers['curtain_module'].settings[0].children[1].type should be equal to one of the allowed values ["text","password","textarea","label"]
manifest.drivers['curtain_module'].settings[0].children[1].step should be number
manifest.drivers['curtain_module'].settings[0].children[1].type should be equal to one of the allowed values ["radio","dropdown"]
manifest.drivers['curtain_module'].settings[0].children[1] should have required property 'values'
manifest.drivers['curtain_module'].settings[0].children[1].type should be equal to one of the allowed values ["checkbox"]
manifest.drivers['curtain_module'].settings[0].children[1].type should be equal to one of the allowed values ["group"]
manifest.drivers['curtain_module'].settings[0].children[1] should have required property 'children'
manifest.drivers['curtain_module'].settings[0].children[1] matched no available schemas, see previous errors
manifest.drivers['curtain_module'].settings[0] matched no available schemas, see previous errors
    at App._validate (C:\Users\ameno\AppData\Roaming\npm\node_modules\homey\lib\App.js:132:13)
    at async App.install (C:\Users\ameno\AppData\Roaming\npm\node_modules\homey\lib\App.js:802:19)
    at async exports.handler (C:\Users\ameno\AppData\Roaming\npm\node_modules\homey\bin\cmds\app\install.js:25:5)
```

**Comment by JohanBendz** [comment]
@RBouroubi Try updating Homey CLI. If this is still an issue please create a separate issue.

**Comment by roxart** [comment]
just installed the new version now my 4 gang does not react at all:
68841fc3-64b9-4c4d-aef6-199fcb634591
started top left 1 click 2 click longpress
top right
bot left
bot right

**Comment by RBouroubi** [comment]
Hi @JohanBendz ,
I've updated my tools, now I could install.

I have now v0.2.72, and the buttons doesn't react at all, no functions.

I've tried to remove and add back one of the devices, same thing, nothing works.

**Comment by JohanBendz** [comment]
Oh.. my bad,, now fixed. Try again @roxart @RBouroubi

**Comment by RBouroubi** [comment]
HI @JohanBendz ,
Seems fine now, I'll do complete tests and get back to you, but feels much better.

**Comment by SunBeech** [comment]
Hi @JohanBendz, I also checked my 4 scene switch and all combinations are working now with a single trigger. Thanks a lot for your efforts!

**Comment by JohanBendz** [comment]
Great to hear @roxart and @SunBeech.
Closing this issue. If you have problems with other devices please start a new issue for that problem.

**Comment by roxart** [comment]
sorry to bother in here - seems to work but I have to additional questions:
how long is the press time for longpress? seems very long and is it adjustable?
2nd is it possible to add "longpress release" card
this to be able to use buttons to do  blinds up and down :) 
also for the lora tap 6 btw ;)

**Comment by SunBeech** [comment]
Hi @roxart, The long press is 5 seconds. That is the time after which the LED indicator light on the button switches off. Also the manual of my MOES device specifies 5 seconds. If you hold it too long (more than 10 seconds), and depending on the button, you enter the pairing mode.

**Comment by roxart** [comment]
ok lets keep this out of here then - and discuss in the forum

**Comment by SunBeech** [comment]
@JohanBendz: issue [#793](https://github.com/JohanBendz/com.tuya.zigbee/issues/793) seems to be the same issue (also for the same device ID). I have been using the work-around mentioned in that issue, but it is no longer necessary as you solved the issue today. I assume that one can also be closed? One additional item off your to do list 😊.
